server/etcdserver/api/v3rpc/qmon.go

```diff
@@ -0,0 +1,272 @@
+	//TODO err handling
+	rss := getCurrentRssBytes()
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	ctrl.resetRespSizeUnsafe()
+	ctrl.updateBudgetUnsafe(uint64(rss))
+	ctrl.updateAuditFlagUnsafe(uint64(rss))
+}
+
+func (ctrl *BandwidthMonitor) resetRespSizeUnsafe() {
+	ctrl.estRespSize, _ = adt.NewWithEstimates(0.0001, 0.9999)
+}
+
+func (ctrl *BandwidthMonitor) updateAuditFlagUnsafe(rss uint64) {
+	ctrl.auditOn = true
+}
+
+func (ctrl *BandwidthMonitor) updateBudgetUnsafe(rss uint64) {
+	if ctrl.totalMemoryBudget <= uint64(rss) {
+		ctrl.budgetExhausted = true
+		runtime.GC()
+	} else {
+		ctrl.budgetExhausted = false
+	}
+}
+
+func (ctrl *BandwidthMonitor) isDeclinedHelper(q Query) (bool, uint64) {
+	respSize := q.qsize
+	if q.qtype == QueryTypeRange {
+		respSize = ctrl.estRespSize.EstimateString(q.qid)
+		if respSize == 0 {
+			//We have not seen a response for this type of query.
+			//Estimate based on default
+			respSize = ctrl.defaultRespSize
+		}
+		q.qsize = respSize
+	}
+	if ctrl.budgetExhausted {
+		//decline
+		return true, respSize
+	}
+
+	//admit
+	return false, respSize
+}
+
+func (ctrl *BandwidthMonitor) newQuery(req interface{}, resp interface{}) Query {
+	var reqSize, respSize int
+	var reqContent string
+	qtype := QueryTypeUnknown
+	switch _resp := resp.(type) {
+	case *pb.RangeResponse:
+		//TODO verify status is 200
+		_req, ok := req.(*pb.RangeRequest)
+		if ok {
+			reqSize = _req.Size()
+			reqContent = _req.String()
+			qtype = QueryTypeRange
+		}
+		if _resp != nil {
+			respSize = _resp.Size()
+		}
+	default:
+		reqSize = 0
+		respSize = 0
+	}
+
+	var q Query
+	q.qid = reqContent
+	q.qsize = uint64(reqSize + respSize)
+	q.qtype = qtype
+	return q
+}
+
+func (ctrl *BandwidthMonitor) newQueryFromReq(req interface{}) Query {
+	var q Query
+	switch req.(type) {
+	case *pb.RangeRequest:
+		_req, ok := req.(*pb.RangeRequest)
+		if ok {
+			q = ctrl.newQRange(_req)
+		}
+	default:
+	}
+	return q
+}
+
+func (ctrl *BandwidthMonitor) newQRange(r *pb.RangeRequest) Query {
+	var q Query
+	q.qid = r.String()
+	q.qsize = 0
+	q.qtype = QueryTypeRange
+	return q
+}
+
+func (ctrl *BandwidthMonitor) UpdateUsage(req interface{}, resp interface{}) {
+	q := ctrl.newQuery(req, resp)
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	ctrl.estRespSize.UpdateString(q.qid, q.qsize)
+	if ctrl.auditOn && q.qtype != QueryTypeUnknown {
+		//TODO use the logger passed from server cfg
+		fmt.Printf("Audit Qid: %s Qsize: %d\n", q.qid, q.qsize)
+	}
+}
+
+func (ctrl *BandwidthMonitor) AdmitReq(req interface{}) bool {
+	q := ctrl.newQueryFromReq(req)
+	declined, qsize := ctrl.isDeclined(q)
+	if qsize > SmallReqThreshold && declined {
+		err := ctrl.throttle.WaitN(context.TODO(), int(qsize))
+		if err != nil {
+			fmt.Printf("Throttled failed. Reject.%s\n", err)
+			return false
+		}
+	}
+	return true
+}
+
+func (ctrl *BandwidthMonitor) isDeclined(q Query) (bool, uint64) {
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	return ctrl.isDeclinedHelper(q)
```

Comment:
isDeclinedUsafe is convention we use in etcd to distinguish locked vs. not-locked logic.

Reply:
ack.  I will update the helper method name